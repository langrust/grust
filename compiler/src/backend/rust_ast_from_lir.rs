//! Rust AST construction from [LIR types](mod@crate::lir).

/// RustAST block construction from LIR block.
pub mod block;
/// RustAST expression construction from LIR expression.
pub mod expression;
/// RustAST item construction from LIR item.
pub mod item;
/// RustAST pattern construction from LIR pattern.
pub mod pattern;
/// RustAST project construction from LIR project.
pub mod project;
/// RustAST statement construction from LIR statement.
pub mod statement;
/// RustAST type construction from LIR type.
pub mod typ;
