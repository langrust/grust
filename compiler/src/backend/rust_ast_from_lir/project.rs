use std::{
    collections::{BTreeMap, BTreeSet},
    path::Path,
};

use crate::{
    backend::rust_ast_from_lir::item::{
        array_alias::rust_ast_from_lir as array_alias_rust_ast_from_lir,
        enumeration::rust_ast_from_lir as enumeration_rust_ast_from_lir,
        function::rust_ast_from_lir as function_rust_ast_from_lir,
        node_file::rust_ast_from_lir as node_file_rust_ast_from_lir,
        structure::rust_ast_from_lir as structure_rust_ast_from_lir,
    },
    lir::{item::Item, project::Project},
};
use itertools::Itertools;
use proc_macro2::Span;
use syn::*;

#[derive(Debug)]
/// Rust project resulting from compiling.
pub struct RustASTProject {
    /// Project's directory.
    pub directory: String,
    files: BTreeMap<String, File>,
    crates: BTreeSet<String>,
}
impl RustASTProject {
    fn new() -> Self {
        RustASTProject {
            files: Default::default(),
            directory: Default::default(),
            crates: Default::default(),
        }
    }

    fn add_file(&mut self, path: &str, file: File) {
        self.files.insert(path.to_owned(), file);
    }

    /// Set project's directory.
    pub fn set_parent<P>(&mut self, path: P)
    where
        P: AsRef<std::path::Path>,
    {
        let subdirectory = std::mem::take(&mut self.directory);
        self.directory = path
            .as_ref()
            .join(subdirectory)
            .into_os_string()
            .into_string()
            .unwrap();
    }
}

/// Transform LIR item into RustAST item.
pub fn rust_ast_from_lir(project: Project) -> RustASTProject {
    let mut rust_ast_project = RustASTProject::new();

    let mut function_file = File {
        shebang: None,
        items: Default::default(),
        attrs: vec![],
    };
    let mut typedefs_file = File {
        shebang: None,
        items: Default::default(),
        attrs: vec![],
    };

    project.items.into_iter().for_each(|item| match item {
        Item::NodeFile(node_file) => {
            let (path, rust_ast_node_file) =
                node_file_rust_ast_from_lir(node_file, &mut rust_ast_project.crates);
            rust_ast_project.add_file(&path, rust_ast_node_file)
        }
        Item::Function(function) => {
            let mut rust_ast_function =
                function_rust_ast_from_lir(function, &mut rust_ast_project.crates);
            function_file.items.append(&mut rust_ast_function);
            // remove duplicated imports between functions
            function_file.items = std::mem::take(&mut function_file.items)
                .into_iter()
                .unique()
                .collect::<Vec<_>>();
        }
        Item::Enumeration(enumeration) => {
            let rust_ast_enumeration = enumeration_rust_ast_from_lir(enumeration);
            typedefs_file
                .items
                .push(syn::Item::Enum(rust_ast_enumeration))
        }
        Item::Structure(structure) => {
            let rust_ast_structure = structure_rust_ast_from_lir(structure);
            typedefs_file
                .items
                .push(syn::Item::Struct(rust_ast_structure))
        }
        Item::ArrayAlias(array_alias) => {
            let rust_ast_array_alias = array_alias_rust_ast_from_lir(array_alias);
            typedefs_file
                .items
                .push(syn::Item::Type(rust_ast_array_alias))
        }
    });

    rust_ast_project.add_file("src/functions.rs", function_file);
    rust_ast_project.add_file("src/typedefs.rs", typedefs_file);

    let mut lib_file = File {
        shebang: None,
        items: Default::default(),
        attrs: vec![],
    };
    rust_ast_project.files.iter().for_each(|(path, _)| {
        let module_name = Path::new(&path)
            .file_stem()
            .unwrap()
            .to_str()
            .unwrap()
            .to_string();
        let module_ident = Ident::new(&module_name, Span::call_site());
        let module_decl = parse_quote! { pub mod #module_ident; };
        lib_file.items.push(syn::Item::Mod(module_decl))
    });
    rust_ast_project.add_file("src/lib.rs", lib_file);

    rust_ast_project
}
