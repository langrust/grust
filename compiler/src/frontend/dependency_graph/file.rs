use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;

prelude! {
    error::{Error, TerminationError},
    hir::file::File,
    symbol_table::SymbolTable,

}

impl File {
    /// Generate dependency graph for every nodes/component.
    pub fn generate_dependency_graphs(
        &mut self,
        symbol_table: &SymbolTable,
        errors: &mut Vec<Error>,
    ) -> Result<(), TerminationError> {
        let File { nodes, .. } = self;

        // initialize dictionariy for reduced graphs
        let mut nodes_reduced_graphs = HashMap::new();

        // create graph of nodes
        let mut nodes_graph = DiGraphMap::new();
        nodes
            .iter()
            .for_each(|node| node.add_node_dependencies(&mut nodes_graph));

        // sort nodes according to their dependencies
        let sorted_nodes = toposort(&nodes_graph, None).map_err(|node| {
            let error = Error::NotCausalNode {
                node: symbol_table.get_name(node.node_id()).clone(),
                location: self.location.clone(),
            };
            errors.push(error);
            TerminationError
        })?;
        nodes.sort_by(|n1, n2| {
            let index1 = sorted_nodes
                .iter()
                .position(|id| *id == n1.id)
                .expect("should be in sorted list");
            let index2 = sorted_nodes
                .iter()
                .position(|id| *id == n2.id)
                .expect("should be in sorted list");

            Ord::cmp(&index2, &index1)
        });

        // ordered nodes complete their dependency graphs
        nodes
            .iter_mut()
            .map(|node| node.compute_dependencies(symbol_table, &mut nodes_reduced_graphs, errors))
            .collect::<Vec<Result<(), TerminationError>>>()
            .into_iter()
            .collect::<Result<(), TerminationError>>()?;

        Ok(())
    }
}
