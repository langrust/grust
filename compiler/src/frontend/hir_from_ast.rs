prelude! {}

/// HIR Component construction from AST Component
pub mod component;
/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
pub mod typ;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST<Ctxt> {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(self, ctxt: &mut Ctxt) -> TRes<Self::HIR>;
}

pub struct SimpleCtxt<'a> {
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct LocCtxt<'a> {
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
pub struct PatLocCtxt<'a> {
    pub pat: Option<&'a ast::stmt::Pattern>,
    pub loc: &'a Location,
    pub syms: &'a mut SymbolTable,
    pub errors: &'a mut Vec<Error>,
}
impl<'a> SimpleCtxt<'a> {
    pub fn new(syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { syms, errors }
    }
    pub fn add_loc<'b>(&'b mut self, loc: &'b Location) -> LocCtxt<'b> {
        LocCtxt::new(loc, self.syms, self.errors)
    }
    pub fn add_pat_loc<'b>(
        &'b mut self,
        pat: Option<&'b ast::stmt::Pattern>,
        loc: &'b Location,
    ) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, loc, self.syms, self.errors)
    }
}
impl<'a> LocCtxt<'a> {
    pub fn new(loc: &'a Location, syms: &'a mut SymbolTable, errors: &'a mut Vec<Error>) -> Self {
        Self { loc, syms, errors }
    }
    pub fn add_pat<'b>(&'b mut self, pat: Option<&'b ast::stmt::Pattern>) -> PatLocCtxt<'b> {
        PatLocCtxt::new(pat, self.loc, self.syms, self.errors)
    }
}
impl<'a> PatLocCtxt<'a> {
    pub fn new(
        pat: Option<&'a ast::stmt::Pattern>,
        loc: &'a Location,
        syms: &'a mut SymbolTable,
        errors: &'a mut Vec<Error>,
    ) -> Self {
        Self {
            pat,
            loc,
            syms,
            errors,
        }
    }
    pub fn remove_pat<'b>(&'b mut self) -> LocCtxt<'b> {
        LocCtxt::new(self.loc, self.syms, self.errors)
    }
    pub fn set_pat(
        &mut self,
        pat: Option<&'a ast::stmt::Pattern>,
    ) -> Option<&'a ast::stmt::Pattern> {
        std::mem::replace(&mut self.pat, pat)
    }
}
