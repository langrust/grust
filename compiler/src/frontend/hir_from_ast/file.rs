prelude! {
    ast::Ast,
    hir::interface::Interface,
}

use super::HIRFromAST;

impl HIRFromAST for Ast {
    type HIR = hir::File;

    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        // initialize symbol table with builtin operators
        symbol_table.initialize();

        // store elements in symbol table
        self.store(symbol_table, errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Function(function) => {
                        functions.push(function.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Typedef(typedef) => {
                        typedefs.push(typedef.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Service(service) => {
                        services.push(service.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Import(import) => imports.push(
                        import
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
