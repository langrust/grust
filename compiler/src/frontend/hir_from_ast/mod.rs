prelude! {}

/// HIR Contract construction from AST Contract
pub mod contract;
/// HIR Equation construction from AST Equation
pub mod equation;
/// HIR Expression construction from AST Expression
pub mod expression;
/// HIR File construction from AST File
pub mod file;
/// HIR Function construction from AST Function
pub mod function;
/// HIR Interface construction from AST Interface.
pub mod interface;
/// HIR Pattern construction from AST Pattern
pub mod pattern;
/// HIR Statement construction from AST Statement
pub mod statement;
/// HIR StreamExpression construction from AST StreamExpression
pub mod stream_expression;
/// HIR Typedef construction from AST Typedef.
pub mod typedef;

/// AST transformation into HIR.
pub trait HIRFromAST {
    /// Corresponding HIR construct.
    type HIR;

    /// Transforms AST into HIR and check identifiers good use.
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR>;
}

impl HIRFromAST for ast::Component {
    type HIR = hir::Node;

    // precondition: node and its signals are already stored in symbol table
    // postcondition: construct HIR node and check identifiers good use
    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        let ast::Component {
            ident,
            contract,
            equations,
            ..
        } = self;
        let name = ident.to_string();
        let location = Location::default();
        let id = symbol_table.get_node_id(&name, false, location.clone(), errors)?;

        // create local context with all signals
        symbol_table.local();
        symbol_table.restore_context(id);
        symbol_table.enter_in_node(id);

        let statements = equations
            .into_iter()
            .map(|equation| equation.hir_from_ast(symbol_table, errors))
            .collect::<TRes<Vec<_>>>()?;
        let contract = contract.hir_from_ast(symbol_table, errors)?;

        symbol_table.leave_node();
        symbol_table.global();

        Ok(hir::Node {
            id,
            statements,
            contract,
            location,
            graph: graph::DiGraphMap::new(),
            memory: hir::Memory::new(),
        })
    }
}
