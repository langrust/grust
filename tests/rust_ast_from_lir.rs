use codespan_reporting::files::{Files, SimpleFiles};

use grustine::ast::file::File;
use grustine::backend::rust_ast_from_lir::project::rust_ast_from_lir;
use grustine::frontend::hir_from_ast::file::hir_from_ast;
use grustine::frontend::lir_from_hir::file::lir_from_hir;
use grustine::parser::langrust;

#[test]
fn rust_ast_from_lir_transformation_for_counter() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/counter.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(counter_id, &files.source(counter_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_blinking() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/blinking.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_button_management() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/button_management.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/button_management_using_function.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}

#[test]
fn rust_ast_from_lir_transformation_for_pid() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/pid.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(pid_id, &files.source(pid_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();
    let mut file = hir_from_ast(file);
    file.generate_dependency_graphs(&mut errors).unwrap();
    file.causality_analysis(&mut errors).unwrap();
    file.normalize(&mut errors).unwrap();
    let project = lir_from_hir(file);

    let project = rust_ast_from_lir(project);
    insta::assert_debug_snapshot!(project);
}
