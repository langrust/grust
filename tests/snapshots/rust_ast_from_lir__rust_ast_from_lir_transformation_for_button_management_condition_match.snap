---
source: tests/rust_ast_from_lir.rs
expression: project
---
directory: ""
files:
  - path: src/counter_o.rs
    items:
      - Structure:
          public_visibility: true
          name: counter_oInput
          fields:
            - public_visibility: true
              name: res
              type:
                Identifier:
                  identifier: bool
            - public_visibility: true
              name: inc
              type:
                Identifier:
                  identifier: i64
      - Structure:
          public_visibility: true
          name: counter_oState
          fields:
            - public_visibility: false
              name: mem_o
              type:
                Identifier:
                  identifier: i64
      - Implementation:
          trait_name: ~
          type_name: counter_oState
          items:
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: init
                  receiver: ~
                  inputs: []
                  output:
                    Identifier:
                      identifier: counter_oState
                body:
                  statements:
                    - ExpressionLast:
                        Structure:
                          name: counter_oState
                          fields:
                            - name: mem_o
                              expression:
                                Literal:
                                  literal:
                                    Integer: 0
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: step
                  receiver:
                    reference: false
                    mutable: false
                  inputs:
                    - - input
                      - Identifier:
                          identifier: counter_oInput
                  output:
                    Tuple:
                      elements:
                        - Identifier:
                            identifier: counter_oState
                        - Identifier:
                            identifier: i64
                body:
                  statements:
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: o
                        expression:
                          IfThenElse:
                            condition:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: input
                                field: res
                            then_branch:
                              statements:
                                - ExpressionLast:
                                    Literal:
                                      literal:
                                        Integer: 0
                            else_branch:
                              statements:
                                - ExpressionLast:
                                    Binary:
                                      left:
                                        Unary:
                                          operator: Brackets
                                          expression:
                                            FieldAccess:
                                              expression:
                                                Identifier:
                                                  identifier: self
                                              field: mem_o
                                      operator: Add
                                      right:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: inc
                    - ExpressionLast:
                        Tuple:
                          elements:
                            - Structure:
                                name: counter_oState
                                fields:
                                  - name: mem_o
                                    expression:
                                      Identifier:
                                        identifier: o
                            - Identifier:
                                identifier: o
  - path: src/reset_button_user_reset_request.rs
    items:
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: typedefs
                    tree:
                      Name:
                        name: Button
                        alias: ~
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: counter_o
                    tree: Star
      - Structure:
          public_visibility: true
          name: reset_button_user_reset_requestInput
          fields:
            - public_visibility: true
              name: button_state
              type:
                Identifier:
                  identifier: Button
            - public_visibility: true
              name: period
              type:
                Identifier:
                  identifier: i64
            - public_visibility: true
              name: reset_limit_1
              type:
                Identifier:
                  identifier: i64
            - public_visibility: true
              name: reset_limit_2
              type:
                Identifier:
                  identifier: i64
      - Structure:
          public_visibility: true
          name: reset_button_user_reset_requestState
          fields:
            - public_visibility: false
              name: mem_res
              type:
                Identifier:
                  identifier: bool
            - public_visibility: false
              name: mem_user_reset_request
              type:
                Identifier:
                  identifier: bool
            - public_visibility: false
              name: counter_o_counter
              type:
                Identifier:
                  identifier: counter_oState
      - Implementation:
          trait_name: ~
          type_name: reset_button_user_reset_requestState
          items:
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: init
                  receiver: ~
                  inputs: []
                  output:
                    Identifier:
                      identifier: reset_button_user_reset_requestState
                body:
                  statements:
                    - ExpressionLast:
                        Structure:
                          name: reset_button_user_reset_requestState
                          fields:
                            - name: mem_res
                              expression:
                                Literal:
                                  literal:
                                    Boolean: true
                            - name: mem_user_reset_request
                              expression:
                                Literal:
                                  literal:
                                    Boolean: false
                            - name: counter_o_counter
                              expression:
                                FunctionCall:
                                  function:
                                    Identifier:
                                      identifier: "counter_oState::init"
                                  arguments: []
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: step
                  receiver:
                    reference: false
                    mutable: false
                  inputs:
                    - - input
                      - Identifier:
                          identifier: reset_button_user_reset_requestInput
                  output:
                    Tuple:
                      elements:
                        - Identifier:
                            identifier: reset_button_user_reset_requestState
                        - Identifier:
                            identifier: bool
                body:
                  statements:
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: res
                        expression:
                          Binary:
                            left:
                              Unary:
                                operator: Brackets
                                expression:
                                  Binary:
                                    left:
                                      FieldAccess:
                                        expression:
                                          Identifier:
                                            identifier: input
                                        field: button_state
                                    operator: Eq
                                    right:
                                      Literal:
                                        literal:
                                          Enumeration:
                                            - Button
                                            - Pressed
                            operator: And
                            right:
                              Unary:
                                operator: Brackets
                                expression:
                                  FieldAccess:
                                    expression:
                                      Identifier:
                                        identifier: self
                                    field: mem_res
                    - Let:
                        pattern:
                          Tuple:
                            elements:
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: counter_o_counter
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: counter
                        expression:
                          MethodCall:
                            receiver:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: self
                                field: counter_o_counter
                            method: step
                            arguments:
                              - Structure:
                                  name: counter_oInput
                                  fields:
                                    - name: res
                                      expression:
                                        Identifier:
                                          identifier: res
                                    - name: inc
                                      expression:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: period
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: user_reset_request
                        expression:
                          Match:
                            matched:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: input
                                field: button_state
                            arms:
                              - pattern:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - Button
                                        - Released
                                guard: ~
                                body:
                                  FieldAccess:
                                    expression:
                                      Identifier:
                                        identifier: self
                                    field: mem_user_reset_request
                              - pattern:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - Button
                                        - Pressed
                                guard: ~
                                body:
                                  Binary:
                                    left:
                                      Identifier:
                                        identifier: counter
                                    operator: Geq
                                    right:
                                      Binary:
                                        left:
                                          FieldAccess:
                                            expression:
                                              Identifier:
                                                identifier: input
                                            field: reset_limit_1
                                        operator: Add
                                        right:
                                          FieldAccess:
                                            expression:
                                              Identifier:
                                                identifier: input
                                            field: reset_limit_2
                    - ExpressionLast:
                        Tuple:
                          elements:
                            - Structure:
                                name: reset_button_user_reset_requestState
                                fields:
                                  - name: mem_res
                                    expression:
                                      Binary:
                                        left:
                                          FieldAccess:
                                            expression:
                                              Identifier:
                                                identifier: input
                                            field: button_state
                                        operator: Eq
                                        right:
                                          Literal:
                                            literal:
                                              Enumeration:
                                                - Button
                                                - Released
                                  - name: mem_user_reset_request
                                    expression:
                                      Identifier:
                                        identifier: user_reset_request
                                  - name: counter_o_counter
                                    expression:
                                      Identifier:
                                        identifier: counter_o_counter
                            - Identifier:
                                identifier: user_reset_request
  - path: src/reset_button_user_reset_request_state.rs
    items:
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: typedefs
                    tree:
                      Name:
                        name: Button
                        alias: ~
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: counter_o
                    tree: Star
      - Import:
          Use:
            public_visibility: false
            tree:
              Path:
                module_name: crate
                tree:
                  Path:
                    module_name: typedefs
                    tree:
                      Name:
                        name: ResetState
                        alias: ~
      - Structure:
          public_visibility: true
          name: reset_button_user_reset_request_stateInput
          fields:
            - public_visibility: true
              name: button_state
              type:
                Identifier:
                  identifier: Button
            - public_visibility: true
              name: period
              type:
                Identifier:
                  identifier: i64
            - public_visibility: true
              name: reset_limit_1
              type:
                Identifier:
                  identifier: i64
      - Structure:
          public_visibility: true
          name: reset_button_user_reset_request_stateState
          fields:
            - public_visibility: false
              name: mem_res
              type:
                Identifier:
                  identifier: bool
            - public_visibility: false
              name: counter_o_counter
              type:
                Identifier:
                  identifier: counter_oState
      - Implementation:
          trait_name: ~
          type_name: reset_button_user_reset_request_stateState
          items:
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: init
                  receiver: ~
                  inputs: []
                  output:
                    Identifier:
                      identifier: reset_button_user_reset_request_stateState
                body:
                  statements:
                    - ExpressionLast:
                        Structure:
                          name: reset_button_user_reset_request_stateState
                          fields:
                            - name: mem_res
                              expression:
                                Literal:
                                  literal:
                                    Boolean: true
                            - name: counter_o_counter
                              expression:
                                FunctionCall:
                                  function:
                                    Identifier:
                                      identifier: "counter_oState::init"
                                  arguments: []
            - AssociatedMethod:
                signature:
                  public_visibility: true
                  name: step
                  receiver:
                    reference: false
                    mutable: false
                  inputs:
                    - - input
                      - Identifier:
                          identifier: reset_button_user_reset_request_stateInput
                  output:
                    Tuple:
                      elements:
                        - Identifier:
                            identifier: reset_button_user_reset_request_stateState
                        - Identifier:
                            identifier: ResetState
                body:
                  statements:
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: res
                        expression:
                          Binary:
                            left:
                              Unary:
                                operator: Brackets
                                expression:
                                  Binary:
                                    left:
                                      FieldAccess:
                                        expression:
                                          Identifier:
                                            identifier: input
                                        field: button_state
                                    operator: Eq
                                    right:
                                      Literal:
                                        literal:
                                          Enumeration:
                                            - Button
                                            - Pressed
                            operator: And
                            right:
                              Unary:
                                operator: Brackets
                                expression:
                                  FieldAccess:
                                    expression:
                                      Identifier:
                                        identifier: self
                                    field: mem_res
                    - Let:
                        pattern:
                          Tuple:
                            elements:
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: counter_o_counter
                              - Identifier:
                                  reference: false
                                  mutable: false
                                  identifier: counter
                        expression:
                          MethodCall:
                            receiver:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: self
                                field: counter_o_counter
                            method: step
                            arguments:
                              - Structure:
                                  name: counter_oInput
                                  fields:
                                    - name: res
                                      expression:
                                        Identifier:
                                          identifier: res
                                    - name: inc
                                      expression:
                                        FieldAccess:
                                          expression:
                                            Identifier:
                                              identifier: input
                                          field: period
                    - Let:
                        pattern:
                          Identifier:
                            reference: false
                            mutable: false
                            identifier: user_reset_request_state
                        expression:
                          Match:
                            matched:
                              FieldAccess:
                                expression:
                                  Identifier:
                                    identifier: input
                                field: button_state
                            arms:
                              - pattern:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - Button
                                        - Released
                                guard: ~
                                body:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - ResetState
                                        - Confirmed
                              - pattern:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - Button
                                        - Pressed
                                guard:
                                  Binary:
                                    left:
                                      Identifier:
                                        identifier: counter
                                    operator: Geq
                                    right:
                                      FieldAccess:
                                        expression:
                                          Identifier:
                                            identifier: input
                                        field: reset_limit_1
                                body:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - ResetState
                                        - InProgress
                              - pattern: Default
                                guard: ~
                                body:
                                  Literal:
                                    literal:
                                      Enumeration:
                                        - ResetState
                                        - Confirmed
                    - ExpressionLast:
                        Tuple:
                          elements:
                            - Structure:
                                name: reset_button_user_reset_request_stateState
                                fields:
                                  - name: mem_res
                                    expression:
                                      Binary:
                                        left:
                                          FieldAccess:
                                            expression:
                                              Identifier:
                                                identifier: input
                                            field: button_state
                                        operator: Eq
                                        right:
                                          Literal:
                                            literal:
                                              Enumeration:
                                                - Button
                                                - Released
                                  - name: counter_o_counter
                                    expression:
                                      Identifier:
                                        identifier: counter_o_counter
                            - Identifier:
                                identifier: user_reset_request_state
  - path: src/functions.rs
    items: []
  - path: src/typedefs.rs
    items:
      - Enumeration:
          public_visibility: true
          name: Button
          elements:
            - Released
            - Pressed
      - Enumeration:
          public_visibility: true
          name: ResetState
          elements:
            - InProgress
            - Confirmed
  - path: src/lib.rs
    items:
      - Import:
          Module:
            public_visibility: true
            name: counter_o
      - Import:
          Module:
            public_visibility: true
            name: reset_button_user_reset_request
      - Import:
          Module:
            public_visibility: true
            name: reset_button_user_reset_request_state
      - Import:
          Module:
            public_visibility: true
            name: functions
      - Import:
          Module:
            public_visibility: true
            name: typedefs

